File: src/backend_dotnet/HRMApi/Services/PointService.cs
using AutoMapper;
using HRMApi.Data;
using HRMApi.DTOs;
using HRMApi.Models;
using HRMApi.Repositories;
using Microsoft.EntityFrameworkCore;

namespace HRMApi.Services;

public class PointService : IPointService
{
    private readonly IPointRepository _pointRepository;
    private readonly IEmployeeRepository _employeeRepository;
    private readonly HrmDbContext _context;
    private readonly IMapper _mapper;
    private readonly ILogger<PointService> _logger;

    public PointService(
        IPointRepository pointRepository,
        IEmployeeRepository employeeRepository,
        HrmDbContext context,
        IMapper mapper,
        ILogger<PointService> logger)
    {
        _pointRepository = pointRepository;
        _employeeRepository = employeeRepository;
        _context = context;
        _mapper = mapper;
        _logger = logger;
    }

    // ============================================
    // POINT QUERIES
    // ============================================
    
    public async Task<ApiResponse<EmployeePointDto>> GetEmployeePointAsync(int employeeId)
    {
        try
        {
            var employeeExists = await _employeeRepository.ExistsAsync(employeeId);
            if (!employeeExists)
            {
                return ApiResponse<EmployeePointDto>.ErrorResponse(
                    "Không tìm thấy nhân viên",
                    new List<string> { $"Nhân viên với ID {employeeId} không tồn tại" });
            }

            var point = await _pointRepository.GetByEmployeeIdAsync(employeeId);
            
            if (point == null)
            {
                return ApiResponse<EmployeePointDto>.ErrorResponse(
                    "Không tìm thấy thông tin điểm",
                    new List<string> { "Nhân viên chưa có thông tin điểm thưởng" });
            }

            var dto = _mapper.Map<EmployeePointDto>(point);
            
            return ApiResponse<EmployeePointDto>.SuccessResponse(
                dto, 
                "Lấy thông tin điểm thành công");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting employee point for {EmployeeId}", employeeId);
            throw;
        }
    }

    public async Task<PagedResult<EmployeePointDto>> GetAllEmployeePointsAsync(
        int pageNumber, 
        int pageSize,
        string? searchTerm = null)
    {
        try
        {
            var (items, totalCount) = await _pointRepository.GetPagedAsync(
                pageNumber, pageSize, searchTerm);

            var dtos = _mapper.Map<List<EmployeePointDto>>(items);

            return new PagedResult<EmployeePointDto>
            {
                Items = dtos,
                TotalCount = totalCount,
                PageNumber = pageNumber,
                PageSize = pageSize
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting all employee points");
            throw;
        }
    }

    // ============================================
    // POINT UPDATES
    // ============================================
    
    public async Task<ApiResponse<EmployeePointDto>> UpdateEmployeePointAsync(
        int employeeId, 
        UpdatePointDto dto)
    {
        await using var transaction = await _context.Database.BeginTransactionAsync();
        
        try
        {
            // Validate employee exists
            var employeeExists = await _employeeRepository.ExistsAsync(employeeId);
            if (!employeeExists)
            {
                return ApiResponse<EmployeePointDto>.ErrorResponse(
                    "Không tìm thấy nhân viên",
                    new List<string> { $"Nhân viên với ID {employeeId} không tồn tại" });
            }

            // Get or create point record
            var point = await _pointRepository.GetByEmployeeIdAsync(employeeId);
            if (point == null)
            {
                return ApiResponse<EmployeePointDto>.ErrorResponse(
                    "Không tìm thấy thông tin điểm",
                    new List<string> { "Nhân viên chưa có thông tin điểm thưởng" });
            }

            // Calculate new point total
            int newTotal = point.PointTotal ?? 0;
            
            switch (dto.Type.ToLower())
            {
                case "earn":
                    newTotal += Math.Abs(dto.Value);
                    break;
                case "redeem":
                    newTotal -= Math.Abs(dto.Value);
                    break;
                case "adjustment":
                    newTotal = dto.Value;
                    break;
            }

            // Validate point total doesn't go negative
            if (newTotal < 0)
            {
                return ApiResponse<EmployeePointDto>.ErrorResponse(
                    "Điểm không hợp lệ",
                    new List<string> { "Tổng điểm không thể âm" });
            }

            // Update point
            point.PointTotal = newTotal;
            point.LastUpdate = DateTime.UtcNow;
            await _pointRepository.UpdateAsync(point);

            // Create transaction history
            var transaction_history = new PointTransactionHistory
            {
                EmployeeId = employeeId,
                Value = dto.Value,
                Type = dto.Type.ToLower(),
                ActorId = dto.ActorId, // ActorId được truyền từ client
                Description = dto.Description ?? $"Cập nhật điểm: {dto.Type}",
                CreatedAt = DateTime.UtcNow
            };
            await _pointRepository.AddTransactionAsync(transaction_history);

            await transaction.CommitAsync();

            // Get updated point with details
            var updatedPoint = await _pointRepository.GetByEmployeeIdAsync(employeeId);
            var resultDto = _mapper.Map<EmployeePointDto>(updatedPoint);

            return ApiResponse<EmployeePointDto>.SuccessResponse(
                resultDto,
                "Cập nhật điểm thành công");
        }
        catch (Exception ex)
        {
            await transaction.RollbackAsync();
            _logger.LogError(ex, "Error updating point for employee {EmployeeId}", employeeId);
            return ApiResponse<EmployeePointDto>.ErrorResponse(
                "Lỗi khi cập nhật điểm",
                new List<string> { ex.Message });
        }
    }

    // ============================================
    // POINT TRANSACTION HISTORY
    // ============================================
    
    public async Task<PagedResult<PointTransactionDto>> GetPointTransactionsAsync(
        int pageNumber,
        int pageSize,
        int? employeeId = null,
        string? type = null,
        DateTime? fromDate = null,
        DateTime? toDate = null)
    {
        try
        {
            var (items, totalCount) = await _pointRepository.GetTransactionsPagedAsync(
                pageNumber, pageSize, employeeId, type, fromDate, toDate);

            var dtos = _mapper.Map<List<PointTransactionDto>>(items);

            return new PagedResult<PointTransactionDto>
            {
                Items = dtos,
                TotalCount = totalCount,
                PageNumber = pageNumber,
                PageSize = pageSize
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting point transactions");
            throw;
        }
    }

    public async Task<List<PointTransactionDto>> GetEmployeeTransactionHistoryAsync(
        int employeeId, 
        int? limit = null)
    {
        try
        {
            var transactions = await _pointRepository.GetTransactionsByEmployeeIdAsync(
                employeeId, limit);

            return _mapper.Map<List<PointTransactionDto>>(transactions);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting transaction history for employee {EmployeeId}", employeeId);
            throw;
        }
    }

    // ============================================
    // POINT CONVERSION RULES
    // ============================================
    
    public async Task<ApiResponse<List<PointConversionRuleDto>>> GetActiveConversionRuleAsync()
    {
        try
        {
            // Gọi hàm repository (trả về IEnumerable)
            var rules = await _pointRepository.GetActiveConversionRulesAsync();
            
            // Nếu null hoặc rỗng
            if (rules == null || !rules.Any())
            {
                return ApiResponse<List<PointConversionRuleDto>>.ErrorResponse(
                    "Không tìm thấy quy tắc quy đổi",
                    new List<string> { "Chưa có quy tắc quy đổi điểm nào được kích hoạt" });
            }

            var dtos = _mapper.Map<List<PointConversionRuleDto>>(rules);
            
            return ApiResponse<List<PointConversionRuleDto>>.SuccessResponse(
                dtos,
                "Lấy danh sách quy tắc quy đổi thành công");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting active conversion rules");
            throw;
        }
    }

    public async Task<List<PointConversionRuleDto>> GetAllConversionRulesAsync()
    {
        try
        {
            var rules = await _pointRepository.GetAllConversionRulesAsync();
            return _mapper.Map<List<PointConversionRuleDto>>(rules);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting all conversion rules");
            throw;
        }
    }

    public async Task<ApiResponse<PointConversionRuleDto>> CreateConversionRuleAsync(
        UpsertPointConversionRuleDto dto)
    {
        await using var transaction = await _context.Database.BeginTransactionAsync();
        
        try
        {
            // Deactivate all existing rules
            var existingRules = await _pointRepository.GetAllConversionRulesAsync();
            foreach (var existingRule in existingRules)
            {
                existingRule.IsActive = false;
                await _pointRepository.UpdateConversionRuleAsync(existingRule);
            }

            // Create new rule
            var rule = new PointConversionRule
            {
                PointValue = dto.PointValue,
                MoneyValue = dto.MoneyValue,
                UpdatedBy = dto.UpdatedBy,
                UpdatedAt = DateTime.UtcNow,
                IsActive = true
            };

            var createdRule = await _pointRepository.AddConversionRuleAsync(rule);
            await transaction.CommitAsync();

            // Get rule with details
            var ruleWithDetails = await _pointRepository.GetConversionRuleByIdAsync(createdRule.Id);
            var resultDto = _mapper.Map<PointConversionRuleDto>(ruleWithDetails);

            return ApiResponse<PointConversionRuleDto>.SuccessResponse(
                resultDto,
                "Tạo quy tắc quy đổi thành công");
        }
        catch (Exception ex)
        {
            await transaction.RollbackAsync();
            _logger.LogError(ex, "Error creating conversion rule");
            return ApiResponse<PointConversionRuleDto>.ErrorResponse(
                "Lỗi khi tạo quy tắc quy đổi",
                new List<string> { ex.Message });
        }
    }

    public async Task<ApiResponse<PointConversionRuleDto>> UpdateConversionRuleAsync(
        int id, 
        UpsertPointConversionRuleDto dto)
    {
        await using var transaction = await _context.Database.BeginTransactionAsync();
        
        try
        {
            var rule = await _pointRepository.GetConversionRuleByIdAsync(id);
            
            if (rule == null)
            {
                return ApiResponse<PointConversionRuleDto>.ErrorResponse(
                    "Không tìm thấy quy tắc",
                    new List<string> { $"Quy tắc với ID {id} không tồn tại" });
            }

            // Deactivate all other rules
            var existingRules = await _pointRepository.GetAllConversionRulesAsync();
            foreach (var existingRule in existingRules.Where(r => r.Id != id))
            {
                existingRule.IsActive = false;
                await _pointRepository.UpdateConversionRuleAsync(existingRule);
            }

            // Update rule
            rule.PointValue = dto.PointValue;
            rule.MoneyValue = dto.MoneyValue;
            rule.UpdatedBy = dto.UpdatedBy;
            rule.UpdatedAt = DateTime.UtcNow;
            rule.IsActive = true;

            await _pointRepository.UpdateConversionRuleAsync(rule);
            await transaction.CommitAsync();

            // Get updated rule with details
            var updatedRule = await _pointRepository.GetConversionRuleByIdAsync(id);
            var resultDto = _mapper.Map<PointConversionRuleDto>(updatedRule);

            return ApiResponse<PointConversionRuleDto>.SuccessResponse(
                resultDto,
                "Cập nhật quy tắc quy đổi thành công");
        }
        catch (Exception ex)
        {
            await transaction.RollbackAsync();
            _logger.LogError(ex, "Error updating conversion rule {RuleId}", id);
            return ApiResponse<PointConversionRuleDto>.ErrorResponse(
                "Lỗi khi cập nhật quy tắc quy đổi",
                new List<string> { ex.Message });
        }
    }

    // ============================================
    // POINT TO MONEY CONVERSION
    // ============================================
    
// backend_dotnet/HRMApi/Services/PointService.cs
// ... existing code ...

    public async Task<ApiResponse<PointToMoneyHistoryDto>> RequestPointToMoneyConversionAsync(
        int employeeId, 
        RequestPointToMoneyDto dto)
    {
        await using var transaction = await _context.Database.BeginTransactionAsync();
        
        try
        {
            // 1. VALIDATION CƠ BẢN
            // Kiểm tra nhân viên tồn tại
            var employeeExists = await _employeeRepository.ExistsAsync(employeeId);
            if (!employeeExists)
            {
                return ApiResponse<PointToMoneyHistoryDto>.ErrorResponse(
                    "Không tìm thấy nhân viên",
                    new List<string> { $"Nhân viên với ID {employeeId} không tồn tại" });
            }

            // [MỚI] Kiểm tra bội số của 100
            if (dto.PointRequested % 100 != 0)
            {
                return ApiResponse<PointToMoneyHistoryDto>.ErrorResponse(
                    "Điểm không hợp lệ",
                    new List<string> { "Số điểm muốn đổi phải là bội số của 100 (Ví dụ: 100, 200, 500...)" });
            }

            // Kiểm tra có request nào đang pending không
            var hasPendingRequest = await _context.PointToMoneyHistories
                .AnyAsync(h => h.EmployeeId == employeeId && h.Status == "pending");
            
            if (hasPendingRequest)
            {
                return ApiResponse<PointToMoneyHistoryDto>.ErrorResponse(
                    "Không thể gửi yêu cầu",
                    new List<string> { "Bạn đang có yêu cầu chờ xử lý. Vui lòng đợi admin duyệt trước." });
            }

            // 2. KIỂM TRA SỐ DƯ ĐIỂM
            var point = await _pointRepository.GetByEmployeeIdAsync(employeeId);
            if (point == null || (point.PointTotal ?? 0) < dto.PointRequested)
            {
                return ApiResponse<PointToMoneyHistoryDto>.ErrorResponse(
                    "Điểm không đủ",
                    new List<string> { $"Bạn chỉ có {point?.PointTotal ?? 0} điểm, không đủ để đổi {dto.PointRequested} điểm." });
            }

            // 3. TÌM QUY TẮC QUY ĐỔI PHÙ HỢP NHẤT
            // Lấy tất cả rule đang active (cần sửa Repository trả về List như hướng dẫn trước)
            var activeRules = await _pointRepository.GetActiveConversionRulesAsync();

            if (activeRules == null || !activeRules.Any())
            {
                return ApiResponse<PointToMoneyHistoryDto>.ErrorResponse(
                    "Lỗi hệ thống",
                    new List<string> { "Chưa có quy tắc quy đổi điểm nào được cấu hình." });
            }

            // Tìm rule có mốc điểm cao nhất mà số điểm yêu cầu thỏa mãn (<= PointRequested)
            // Ví dụ: Đổi 600 điểm. Có mốc 500 và 100. Sẽ chọn mốc 500 để tính tỉ giá.
            var matchedRule = activeRules
                .Where(r => r.PointValue <= dto.PointRequested)
                .OrderByDescending(r => r.PointValue)
                .FirstOrDefault();

            if (matchedRule == null)
            {
                return ApiResponse<PointToMoneyHistoryDto>.ErrorResponse(
                    "Chưa đủ điểm tối thiểu",
                    new List<string> { $"Số điểm tối thiểu để quy đổi là {activeRules.Min(r => r.PointValue)} điểm." });
            }

            // 4. TÍNH TIỀN
            // Công thức: (Điểm yêu cầu / Mốc điểm rule) * Giá tiền rule
            // Lưu ý ép kiểu (decimal) để tránh phép chia số nguyên bị làm tròn sai
            decimal moneyReceived = ((decimal)dto.PointRequested / matchedRule.PointValue) * matchedRule.MoneyValue;

            // 5. LƯU VÀO DB
            var history = new PointToMoneyHistory
            {
                EmployeeId = employeeId,
                PointRequested = dto.PointRequested,
                MoneyReceived = moneyReceived,
                Status = "pending",
                CreatedAt = DateTime.UtcNow
            };

            var createdHistory = await _pointRepository.AddPointToMoneyHistoryAsync(history);
            await transaction.CommitAsync();

            // 6. TRẢ KẾT QUẢ
            var historyWithDetails = await _pointRepository.GetPointToMoneyHistoryByIdAsync(createdHistory.Id);
            var resultDto = _mapper.Map<PointToMoneyHistoryDto>(historyWithDetails);

            return ApiResponse<PointToMoneyHistoryDto>.SuccessResponse(
                resultDto,
                "Gửi yêu cầu quy đổi điểm thành công");
        }
        catch (Exception ex)
        {
            await transaction.RollbackAsync();
            _logger.LogError(ex, "Error requesting point conversion for employee {EmployeeId}", employeeId);
            return ApiResponse<PointToMoneyHistoryDto>.ErrorResponse(
                "Lỗi khi gửi yêu cầu quy đổi",
                new List<string> { ex.Message });
        }
    }

    public async Task<PagedResult<PointToMoneyHistoryDto>> GetPointToMoneyHistoryAsync(
        int pageNumber,
        int pageSize,
        int? employeeId = null,
        string? status = null)
    {
        try
        {
            var (items, totalCount) = await _pointRepository.GetPointToMoneyHistoryPagedAsync(
                pageNumber, pageSize, employeeId, status);

            var dtos = _mapper.Map<List<PointToMoneyHistoryDto>>(items);

            return new PagedResult<PointToMoneyHistoryDto>
            {
                Items = dtos,
                TotalCount = totalCount,
                PageNumber = pageNumber,
                PageSize = pageSize
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting point to money history");
            throw;
        }
    }

    public async Task<ApiResponse<PointToMoneyHistoryDto>> ProcessPointToMoneyRequestAsync(
        int requestId, 
        ProcessPointToMoneyDto dto,
        int? processorId = null)
    {
        await using var transaction = await _context.Database.BeginTransactionAsync();
        
        try
        {
            // Get request
            var history = await _pointRepository.GetPointToMoneyHistoryByIdAsync(requestId);
            
            if (history == null)
            {
                return ApiResponse<PointToMoneyHistoryDto>.ErrorResponse(
                    "Không tìm thấy yêu cầu",
                    new List<string> { $"Yêu cầu với ID {requestId} không tồn tại" });
            }

            if (history.Status != "pending")
            {
                return ApiResponse<PointToMoneyHistoryDto>.ErrorResponse(
                    "Yêu cầu đã được xử lý",
                    new List<string> { "Không thể xử lý yêu cầu đã được duyệt hoặc từ chối" });
            }

            // If approved, deduct points
            if (dto.Status == "approved")
            {
                var point = await _pointRepository.GetByEmployeeIdAsync(history.EmployeeId);
                
                if (point == null || (point.PointTotal ?? 0) < history.PointRequested)
                {
                    return ApiResponse<PointToMoneyHistoryDto>.ErrorResponse(
                        "Điểm không đủ",
                        new List<string> { "Nhân viên không còn đủ điểm để quy đổi" });
                }

                point.PointTotal = (point.PointTotal ?? 0) - history.PointRequested;
                point.LastUpdate = DateTime.UtcNow;
                await _pointRepository.UpdateAsync(point);

                // Create transaction history
                var transactionHistory = new PointTransactionHistory
                {
                    EmployeeId = history.EmployeeId,
                    Value = -history.PointRequested,
                    Type = "redeem",
                    ActorId = processorId,
                    Description = $"Quy đổi {history.PointRequested} điểm sang {history.MoneyReceived:N0} VNĐ",
                    CreatedAt = DateTime.UtcNow
                };
                await _pointRepository.AddTransactionAsync(transactionHistory);
            }

            // Update history status
            history.Status = dto.Status;
            history.ProcessedAt = DateTime.UtcNow;
            await _pointRepository.UpdatePointToMoneyHistoryAsync(history);

            await transaction.CommitAsync();

            // Get updated history with details
            var updatedHistory = await _pointRepository.GetPointToMoneyHistoryByIdAsync(requestId);
            var resultDto = _mapper.Map<PointToMoneyHistoryDto>(updatedHistory);

            return ApiResponse<PointToMoneyHistoryDto>.SuccessResponse(
                resultDto,
                dto.Status == "approved" ? "Duyệt yêu cầu thành công" : "Từ chối yêu cầu thành công");
        }
        catch (Exception ex)
        {
            await transaction.RollbackAsync();
            _logger.LogError(ex, "Error processing point conversion request {RequestId}", requestId);
            return ApiResponse<PointToMoneyHistoryDto>.ErrorResponse(
                "Lỗi khi xử lý yêu cầu",
                new List<string> { ex.Message });
        }
    }

    // ============================================
    // STATISTICS
    // ============================================
    
    public async Task<PointStatisticsDto> GetPointStatisticsAsync()
    {
        try
        {
            var totalEmployees = await _pointRepository.GetTotalEmployeesWithPointsAsync();
            var totalPoints = await _pointRepository.GetTotalPointsInSystemAsync();
            var topEmployee = await _pointRepository.GetTopEmployeeByPointsAsync();
            var pendingRequests = await _pointRepository.GetPendingConversionRequestsCountAsync();
            var transactionsByType = await _pointRepository.GetTransactionCountsByTypeAsync();

            return new PointStatisticsDto
            {
                TotalEmployeesWithPoints = totalEmployees,
                TotalPointsInSystem = totalPoints,
                AveragePointsPerEmployee = totalEmployees > 0 ? (double)totalPoints / totalEmployees : 0,
                TopEmployee = topEmployee != null ? _mapper.Map<EmployeePointDto>(topEmployee) : null,
                PendingConversionRequests = pendingRequests,
                TransactionsByType = transactionsByType
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting point statistics");
            throw;
        }
    }
}

File: src/frontend/src/services/pointService.ts
/**
 * pointService.ts - Service cho quản lý điểm thưởng
 * API calls: xem điểm, lịch sử giao dịch, quy đổi điểm
 */

import type { Update } from '@reduxjs/toolkit';
import { apiDotNet } from './api';

// ============================================
// TYPES & INTERFACES
// ============================================

export interface EmployeePointDto {
  id: number;
  employeeId: number;
  employeeName: string;
  email: string;
  pointTotal: number;
  lastUpdate: string;
}

export interface PointTransactionDto {
  id: number;
  employeeId: number;
  employeeName: string;
  value: number;
  type: string;
  typeDisplay: string;
  actorId: number | null;
  actorName: string | null;
  description: string | null;
  createdAt: string;
}

export interface PointConversionRuleDto {
  id: number;
  pointValue: number;
  moneyValue: number;
  updatedBy: number | null;
  updatedByName: string | null;
  updatedAt: string;
  isActive: boolean;
}

export interface PointToMoneyHistoryDto {
  id: number;
  employeeId: number;
  employeeName: string;
  email: string;
  pointRequested: number;
  moneyReceived: number;
  status: string;
  statusDisplay: string;
  createdAt: string;
  processedAt: string | null;
}

export interface ApiResponse<T> {
  success: boolean;
  message: string;
  data: T;
  errors: string[];
}

export interface PagedResult<T> {
  items: T[];
  totalCount: number;
  pageNumber: number;
  pageSize: number;
  totalPages: number;
  hasPreviousPage: boolean;
  hasNextPage: boolean;
}

export interface MonthlyPointRuleDto {
  id: number;
  roleId: number;
  roleName: string;
  pointValue: number;
  createdAt: string;
  updatedAt: string;
}

export interface UpsertMonthlyPointRuleDto {
  roleId: number;
  pointValue: number;
}

// [NEW] Interface cho History Allocation
export interface MonthlyPointAllocationHistoryDto {
  id: number;
  month: number;
  year: number;
  totalPointsDistributed: number;
  employeeCount: number;
  status: string; // 'Success', 'Failed'
  createdAt: string;
}

export interface MonthlyPointAllocationResultDto {
  success: boolean;
  message: string;
  totalEmployees: number;
  totalPoints: number;
}

export interface UpsertPointConversionRuleDto {
  pointValue: number;
  moneyValue: number;
  isActive: boolean;
}
// ============================================
// SERVICE FUNCTIONS
// ============================================

export const pointService = {
  /**
   * Lấy thông tin điểm thưởng của nhân viên
   * GET /api/Point/employee/{employeeId}
   */
  getEmployeePoint: async (employeeId: number): Promise<EmployeePointDto> => {
    const response = await apiDotNet.get<ApiResponse<EmployeePointDto>>(
      `/Point/employee/${employeeId}`
    );

    if (response.data.success) {
      return response.data.data;
    }

    throw new Error(response.data.message || 'Lỗi khi lấy thông tin điểm');
  },

  /**
   * Lấy lịch sử giao dịch điểm của nhân viên
   * GET /api/Point/transactions/employee/{employeeId}
   */
  getEmployeeTransactionHistory: async (
    employeeId: number,
    limit?: number
  ): Promise<PointTransactionDto[]> => {
    const response = await apiDotNet.get<ApiResponse<PointTransactionDto[]>>(
      `/Point/transactions/employee/${employeeId}`,
      { params: { limit } }
    );

    if (response.data.success) {
      return response.data.data;
    }

    throw new Error(response.data.message || 'Lỗi khi lấy lịch sử giao dịch');
  },

  /**
   * Lấy lịch sử giao dịch điểm với phân trang
   * GET /api/Point/transactions
   */
  getPointTransactions: async (
    pageNumber: number = 1,
    pageSize: number = 10,
    employeeId?: number,
    type?: string,
    fromDate?: string,
    toDate?: string
  ): Promise<PagedResult<PointTransactionDto>> => {
    const response = await apiDotNet.get<PagedResult<PointTransactionDto>>(
      '/Point/transactions',
      {
        params: {
          pageNumber,
          pageSize,
          employeeId,
          type,
          fromDate,
          toDate,
        },
      }
    );

    return response.data;
  },

  /**
   * Lấy quy tắc quy đổi điểm đang hoạt động
   * GET /api/Point/conversion-rules/active
   */
  getActiveConversionRule: async (): Promise<PointConversionRuleDto> => {
    const response = await apiDotNet.get<ApiResponse<PointConversionRuleDto>>(
      '/Point/conversion-rules/active'
    );

    if (response.data.success) {
      return response.data.data;
    }

    throw new Error(response.data.message || 'Không tìm thấy quy tắc quy đổi');
  },

  /**
   * Lấy toàn bộ quy tắc quy đổi điểm
   * GET /api/Point/conversion-rules
   */
  getAllConversionRule: async (): Promise<PointConversionRuleDto[]> => {
    const response = await apiDotNet.get<ApiResponse<PointConversionRuleDto[]>>(
      '/Point/conversion-rules'
    )

    if (response.data.success) {
      return response.data.data;
    }

    throw new Error(response.data.message || 'Không tìm thấy bất kì quy tắc quy đổi');
  },

  /**
   * Gửi yêu cầu quy đổi điểm sang tiền
   * POST /api/Point/employee/{employeeId}/convert
   */
  requestPointToMoneyConversion: async (
    employeeId: number,
    pointRequested: number
  ): Promise<PointToMoneyHistoryDto> => {
    const response = await apiDotNet.post<ApiResponse<PointToMoneyHistoryDto>>(
      `/Point/employee/${employeeId}/convert`,
      { PointRequested: pointRequested }  // Backend .NET expects PascalCase
    );

    if (response.data.success) {
      return response.data.data;
    }

    throw new Error(response.data.message || 'Lỗi khi gửi yêu cầu quy đổi');
  },

  /**
   * Lấy lịch sử quy đổi điểm sang tiền
   * GET /api/Point/conversion-history
   */
  getPointToMoneyHistory: async (
    pageNumber: number = 1,
    pageSize: number = 10,
    employeeId?: number,
    status?: string
  ): Promise<PagedResult<PointToMoneyHistoryDto>> => {
    const response = await apiDotNet.get<PagedResult<PointToMoneyHistoryDto>>(
      '/Point/conversion-history',
      {
        params: {
          pageNumber,
          pageSize,
          employeeId,
          status,
        },
      }
    );

    return response.data;
  },

  /**
   * Lấy thống kê điểm thưởng
   * GET /api/Point/statistics
   */
  getPointStatistics: async (): Promise<any> => {
    const response = await apiDotNet.get<ApiResponse<any>>('/Point/statistics');

    if (response.data.success) {
      return response.data.data;
    }

    throw new Error(response.data.message || 'Lỗi khi lấy thống kê');
  },

  /**
   * Lấy danh sách điểm của tất cả nhân viên
   * GET /api/Point/employees
   */
  getAllEmployeePoints: async (
    pageNumber: number = 1,
    pageSize: number = 100
  ): Promise<PagedResult<EmployeePointDto>> => {
    const response = await apiDotNet.get<PagedResult<EmployeePointDto>>(
      '/Point',
      {
        params: {
          pageNumber,
          pageSize,
        },
      }
    );

    return response.data;
  },
  getMonthlyPointRules: async (): Promise<MonthlyPointRuleDto[]> => {
    const response = await apiDotNet.get<ApiResponse<MonthlyPointRuleDto[]>>(
      '/MonthlyPoint/rules'
    );
    return response.data.data;
  },

  /**
   * Tạo hoặc cập nhật quy tắc cộng điểm
   * POST /api/MonthlyPoint/rules
   */
  upsertMonthlyPointRule: async (dto: UpsertMonthlyPointRuleDto): Promise<MonthlyPointRuleDto> => {
    const response = await apiDotNet.post<ApiResponse<MonthlyPointRuleDto>>(
      '/MonthlyPoint/rules',
      dto
    );
    if (!response.data.success) {
      throw new Error(response.data.message);
    }
    return response.data.data;
  },

  /**
   * Xóa quy tắc cộng điểm
   * DELETE /api/MonthlyPoint/rules/{id}
   */
  deleteMonthlyPointRule: async (id: number): Promise<boolean> => {
    const response = await apiDotNet.delete<ApiResponse<boolean>>(
      `/MonthlyPoint/rules/${id}`
    );
    return response.data.success;
  },

  /**
   * Chạy phân phối điểm thủ công
   * POST /api/MonthlyPoint/allocate
   */
  allocateMonthlyPoints: async (): Promise<MonthlyPointAllocationResultDto> => {
    const response = await apiDotNet.post<ApiResponse<MonthlyPointAllocationResultDto>>(
      '/MonthlyPoint/allocate',
      {}
    );
    if (!response.data.success) {
      throw new Error(response.data.message);
    }
    return response.data.data;
  },

  /**
   * Lấy lịch sử phân phối điểm
   * GET /api/MonthlyPoint/history
   */
  updateConversionRule: async (id: number, dto: UpsertPointConversionRuleDto): Promise<PointConversionRuleDto> => {
    const response = await apiDotNet.put<ApiResponse<PointConversionRuleDto>>(
      `/Point/conversion-rules/${id}`,
      dto
    );
    
    if (response.data.success) {
      return response.data.data;
    }
    throw new Error(response.data.message || 'Lỗi khi cập nhật quy tắc');
  },

  getAllocationHistory: async (limit: number = 12): Promise<MonthlyPointAllocationHistoryDto[]> => {
    const response = await apiDotNet.get<ApiResponse<MonthlyPointAllocationHistoryDto[]>>(
      '/MonthlyPoint/history',
      { params: { limit } }
    );
    return response.data.data;
  },

  getPointToMoneyRequest: async (
    pageNumber: number, 
    pageSize: number, 
    employeeId?: number, 
    status?: string
  ): Promise<PagedResult<PointToMoneyHistoryDto>> => {
    const params: any = { pageNumber, pageSize };
    if (employeeId) params.employeeId = employeeId;
    if (status) params.status = status;

    const response = await apiDotNet.get<PagedResult<PointToMoneyHistoryDto>>(
      '/Point/conversion-history',
      { params }
    );
    return response.data; // Backend trả về trực tiếp PagedResult, không bọc ApiResponse ở lớp ngoài cùng cho GET list
  },

  /**
   * Duyệt hoặc Từ chối yêu cầu đổi điểm
   * PUT /api/v1/Point/conversion-history/{requestId}/process
   */
  processConversionRequest: async (requestId: number, status: 'approved' | 'rejected'): Promise<boolean> => {
    const response = await apiDotNet.put<ApiResponse<any>>(
      `/Point/conversion-history/${requestId}/process`,
      { status }
    );
    
    if (response.data.success) {
      return true;
    }
    throw new Error(response.data.message || "Lỗi khi xử lý yêu cầu");
  },
};